File: src/pkg01_topic_cpp/src/publisher_node.cpp
// publisher_node.cpp
// A minimal ROS2 publisher in C++.
// Publishes a counter as std_msgs/String on the topic "/chatter" at 2 Hz.

#include <chrono>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

using namespace std::chrono_literals;

class PublisherNode : public rclcpp::Node
{
public:
  PublisherNode()
  : Node("publisher_node"), count_(0)
  {
    // Create a publisher:
    //   topic name   : "chatter"
    //   message type : std_msgs/msg/String
    //   QoS depth    : 10 (history buffer size)
    publisher_ = this->create_publisher<std_msgs::msg::String>("chatter", 10);

    // A wall timer fires every 500 ms and calls publish_message().
    timer_ = this->create_wall_timer(
      500ms, std::bind(&PublisherNode::publish_message, this));

    RCLCPP_INFO(this->get_logger(), "Publisher node started, publishing on /chatter");
  }

private:
  void publish_message()
  {
    auto message = std_msgs::msg::String();
    message.data = "Hello ROS2 from C++ #" + std::to_string(count_++);

    RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
    publisher_->publish(message);
  }

  rclcpp::TimerBase::SharedPtr timer_;
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
  size_t count_;
};

int main(int argc, char * argv[])
{
  // Initialize the ROS2 client library
  rclcpp::init(argc, argv);

  // Spin keeps the node alive and processes callbacks
  rclcpp::spin(std::make_shared<PublisherNode>());

  // Cleanup
  rclcpp::shutdown();
  return 0;
}

File: src/pkg02_topic_py/pkg02_topic_py/publisher_node.py
#!/usr/bin/env python3
"""
A minimal ROS2 publisher in Python.
Publishes a counter as std_msgs/String on the topic /chatter at 2 Hz.
"""

import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class PublisherNode(Node):
    def __init__(self):
        super().__init__('publisher_node_py')

        # Create publisher:
        #   message type : String
        #   topic name   : 'chatter'
        #   QoS depth    : 10
        self.publisher_ = self.create_publisher(String, 'chatter', 10)

        # Timer fires every 0.5 s and calls timer_callback
        self.timer = self.create_timer(0.5, self.timer_callback)
        self.count = 0

        self.get_logger().info('Publisher node (Python) started on /chatter')

    def timer_callback(self):
        msg = String()
        msg.data = f'Hello ROS2 from Python #{self.count}'
        self.publisher_.publish(msg)
        self.get_logger().info(f'Publishing: "{msg.data}"')
        self.count += 1


def main(args=None):
    rclpy.init(args=args)            # 1. Init
    node = PublisherNode()
    try:
        rclpy.spin(node)             # 2. Spin (blocks)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()             # 3. Shutdown


if __name__ == '__main__':
    main()
